### PR TITLE
scale_factor configurable for cudnn attention. 

### PR DESCRIPTION
**Pull request by suexu1025**
# Description
Duplicate of [PR #1553 ]. That PR stuck at commit updating, recreate new one to complete merge.

GPU correctness tests are failing due to an issue with fused query weights. The fix is to set their scale_factor to 1.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/404877511

# Tests

1. verified with llama3.1 405b checkpoint with correct eval loss.
2. will add MaxText/tests/forward_pass_logit_checker.py for llama2-70b on GPU on later PR.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/attentions.py [hunk 1]
```diff
@@ -273,6 +273,8 @@ class AttentionOp(nn.Module):
   use_ragged_attention: bool = False
   ragged_block_size: int = 256
 
+  scale_factor: float = 1  # scaling factor for query in attention. currently used in cudnn only.
```
Comment (gagika) [inline MaxText/layers/attentions.py:276]: let's not have an extra config unless there is a strong reason for that, we already have a way to scale query, e.g. `query_pre_attn_scalar` which has the same effect on the attention.

File: MaxText/layers/attentions.py [hunk 2]
```diff
@@ -632,6 +634,12 @@ def cudnn_flash_attention(
       mask_type = "padding_causal"  # only padding_causal mask type can take a created mask
       attn_mask = self.generate_attention_mask(query, key, decoder_segment_ids, model_mode)
 
+    # scaling factor for fused query weight or not
+    if self.scale_factor is None:
+      scale_factor = 1.0 / jnp.sqrt(query.shape[-1])
+    else:
+      scale_factor = self.scale_factor
+
```
Comment (gagika) [inline MaxText/layers/attentions.py:642]: same comment as above, let's just use 1 to be consistent with the TPU and DotProduct attention.

**Review by yangyuwei — APPROVED** [review]
Looks great! Thank you.

File: MaxText/layers/attentions.py (hunk 1 above)
Comment (suexu1025) [reply MaxText/layers/attentions.py]: reused the `query_pre_attn_scalar` for the scaling

File: MaxText/layers/attentions.py [hunk 3]
```diff
@@ -643,7 +651,7 @@ def cudnn_flash_attention(
         dtype=self.dtype,
         float32_logits=self.float32_logits,
         qkv_layout="BSHD_BSHD_BSHD",  # 'BS3HD', 'BSHD_BS2HD' or 'BSHD_BSHD_BSHD'
-        scale_factor=1.0 / math.sqrt(head_dim),
+        scale_factor=query_pre_attn_scalar,
```
Comment (gagika) [inline MaxText/layers/attentions.py:654]: config query_pre_attn_scalar is already do scaling, we can pass 1 here.

File: MaxText/layers/attentions.py [hunk 4]
```diff
@@ -632,6 +634,12 @@ def cudnn_flash_attention(
       mask_type = "padding_causal"  # only padding_causal mask type can take a created mask
       attn_mask = self.generate_attention_mask(query, key, decoder_segment_ids, model_mode)
 
+    # scaling factor for fused query weight or not
+    if self.query_pre_attn_scalar is None:
+      query_pre_attn_scalar = 1.0 / jnp.sqrt(query.shape[-1])
+    else:
+      query_pre_attn_scalar = self.query_pre_attn_scalar
```
Comment (gagika) [inline MaxText/layers/attentions.py:641]: I don't think this is needed, as config.query_pre_attn_scalar is already used for scaling.

**Review by gagika — COMMENTED** [review]
we can just keep `scale_factor=1` to be consistent across accelerates.

File: end_to_end/gpu/a3/test_gemma3_logits.sh [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+# export DATASET_PATH=gs://maxtext-dataset
+# # Non-Googlers please remember to point `BASE_OUTPUT_DIRECTORY` to a GCS bucket that you own, this bucket will store all the files generated by MaxText during a run
+# export BASE_OUTPUT_DIRECTORY=gs://runner-maxtext-logs
+# # We define `CONVERTED_CHECKPOINT` to refer to the checkpoint subdirectory. This way it is easier to use this path in the `train.py` and `decode.py` commands
+# export CONVERTED_CHECKPOINT=${MODEL_BUCKET}/${MODEL_VARIATION}/${idx}/0/items
+# export RUN_NAME=unscanned_chkpt_${idx}
+# export UNSCANNED_CKPT_PATH=${BASE_OUTPUT_DIRECTORY}/${RUN_NAME}/checkpoints/0/items
+# Note that the `CONVERTED_CHECKPOINT` is in a `scanned` format which is great for training but for efficient decoding performance we want the checkpoint in an `unscanned` format.
+# We can do this by running `MaxText/generate_param_only_checkpoint.py` on `CONVERTED_CHECKPOINT` with `force_unroll=true`.
+#JAX_PLATFORMS=cpu python3 -m MaxText.generate_param_only_checkpoint MaxText/configs/base.yml base_output_directory=${BASE_OUTPUT_DIRECTORY} load_parameters_path=${CONVERTED_CHECKPOINT} run_name=${RUN_NAME} model_name=${MODEL_NAME} force_unroll=true
```
Comment (gagika) [inline end_to_end/gpu/a3/test_gemma3_logits.sh:42]: do you need this?

Comment (suexu1025) [reply end_to_end/gpu/a3/test_gemma3_logits.sh:42]: This is the step of how I generate the unscaned ckpt on tpu. 
On gpu, this step generating errors. so I commented those steps out. wdyt?

Comment (gagika) [reply end_to_end/gpu/a3/test_gemma3_logits.sh:42]: will they generate errors if you run with with `JAX_PLATFORMS=cpu`, also will end_to_end/gpu/a3/test_gemma3_logits.sh be added in nightly tests?

Comment (suexu1025) [reply end_to_end/gpu/a3/test_gemma3_logits.sh:42]: yes, it will generate errors as we specify attention=cudnn_flash_te.

**Review by gagika — APPROVED** [review]
Thanks for the fix!
